src/main/kotlin/org/rust/lang/core/resolve/indexes/RustModulesIndex.kt

```diff
@@ -1,40 +1,46 @@
 package org.rust.lang.core.resolve.indexes
 
-import com.intellij.openapi.module.Module
+import com.intellij.openapi.util.io.FileUtil
 import com.intellij.psi.search.GlobalSearchScope
 import com.intellij.util.indexing.FileBasedIndex
 import com.intellij.util.indexing.ID
-import org.rust.lang.core.names.RustQualifiedName
+import org.rust.cargo.util.getPsiFor
 import org.rust.lang.core.psi.RustMod
 import org.rust.lang.core.psi.impl.RustFile
-import org.rust.lang.core.psi.impl.cratePath
-import org.rust.lang.core.psi.util.module
-import org.rust.lang.core.resolve.RustResolveEngine
-
-interface RustModulesIndex {
-
-    companion object {
-
-        val ID: ID<RustCratePath, RustQualifiedName> =
-            com.intellij.util.indexing.ID.create("org.rust.lang.indexes.RustModulesIndex")
-
-        fun getSuperFor(mod: RustFile): RustMod? =
-            mod.containingFile.originalFile.let { file ->
-                mod.module?.let { module ->
-                    file.cratePath?.let { path ->
-                        findByHeterogeneous(
-                            FileBasedIndex.getInstance()
-                                .getValues(ID, path, GlobalSearchScope.allScope(module.project))
-                                .firstOrNull(),
-                            module
-                        )
-                    }
-                }
-            }
+import org.rust.lang.core.psi.util.parentOfType
+
+object RustModulesIndex {
+    val ID: ID<RustModulesIndexExtension.Key, RustModulesIndexExtension.Value> =
+        com.intellij.util.indexing.ID.create("org.rust.lang.indexes.RustModulesIndex")
+
+    fun getSuperFor(mod: RustFile): RustMod? {
+        val project = mod.project
+        val key = getKey(mod) ?: return null
+
+        var result: RustMod? = null
+        FileBasedIndex.getInstance().processValues(ID, key, null, { file, value ->
+            val reference = project.getPsiFor(file)?.findReferenceAt(value.referenceOffset)
+                ?: return@processValues true
 
-        private fun findByHeterogeneous(path: RustQualifiedName?, module: Module): RustMod? =
-            path?.let {
-                RustResolveEngine.resolve(path, module).resolved as RustMod?
+            if (reference.resolve() == mod.originalFile) {
```

Comment:
The reference at offset is guaranteed to be a mod_decl (`mod foo;`). `mod_decl` resolving is [non recursive] and does not and should not call any other resolve routines. 

I think that that "reference is mod_decl" invariant can be made explicit by switching to stub based index instead of file based index. I believe it is possible to implement this index on top of stubs, but I'd rather use a simpler file based approach first. 

[non recursive]: https://github.com/intellij-rust/intellij-rust/blob/master/src/main/kotlin/org/rust/lang/core/resolve/RustResolveEngine.kt#L92-L111